aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt

```diff
@@ -802,15 +842,115 @@ internal class RealAWSCognitoAuthPlugin(
     ) {
-        TODO("Not yet implemented")
+        authStateMachine.getCurrentState { authState ->
+            when (authState.authNState) {
+                // Check if user signed in
+                is AuthenticationState.SignedIn -> {
+                    GlobalScope.launch {
+                        try {
+                            val userAttributesOptions = options as? AWSCognitoAuthUpdateUserAttributesOptions
+                            onSuccess.accept(updateUserAttributes(attributes, userAttributesOptions))
+                        } catch (e: Exception) {
+                            onError.accept(CognitoAuthExceptionConverter.lookup(e, e.toString()))
+                        }
+                    }
+                }
+                else -> onError.accept(AuthException.InvalidStateException())
+            }
+        }
     }
 
     override fun updateUserAttributes(
         attributes: MutableList<AuthUserAttribute>,
         onSuccess: Consumer<MutableMap<AuthUserAttributeKey, AuthUpdateAttributeResult>>,
         onError: Consumer<AuthException>
     ) {
-        TODO("Not yet implemented")
+        authStateMachine.getCurrentState { authState ->
+            when (authState.authNState) {
+                // Check if user signed in
+                is AuthenticationState.SignedIn -> {
+                    GlobalScope.launch {
+                        try {
+                            onSuccess.accept(updateUserAttributes(attributes, null))
+                        } catch (e: Exception) {
+                            onError.accept(CognitoAuthExceptionConverter.lookup(e, e.toString()))
+                        }
+                    }
+                }
+                else -> onError.accept(AuthException.InvalidStateException())
+            }
+        }
+    }
+
+    private suspend fun updateUserAttributes(
+        attributes: MutableList<AuthUserAttribute>,
+        userAttributesOptions: AWSCognitoAuthUpdateUserAttributesOptions?
+    ):
+        MutableMap<AuthUserAttributeKey, AuthUpdateAttributeResult> {
+            val accessToken = getSession().userPoolTokens.value?.accessToken
+            var userAttributes = attributes.map {
+                AttributeType.invoke {
+                    name = it.key.keyString
+                    value = it.value
+                }
+            }
+            val userAttributesRequest = UpdateUserAttributesRequest.invoke {
+                this.accessToken = accessToken
+                this.userAttributes = userAttributes
+                this.clientMetadata = userAttributesOptions?.metadata
+            }
+            val userAttributeResponse = authEnvironment.cognitoAuthService
+                .cognitoIdentityProviderClient?.updateUserAttributes(
+                    userAttributesRequest
+                )
+            return getUpdateUserAttributeResult(userAttributeResponse, userAttributes)
+        }
+
+    private fun getUpdateUserAttributeResult(
```

Comment:
Looks like this mapper itself needs some testing to be covered. 
Although overall code can be simplified more, I think it is important to have confidence on this mapping functions behavior.

Reply:
I checked and the response only returns attributes that require code delivery so this logic seem about right.